Fix conflict about getPortByIp

FengyunPan opened the pull request:
**What this PR does / why we need it**:
Currently getPortByIp() get port of instance only based on IP.
If there are two instances in diffent network and the CIDR of
their subnet are same, getPortByIp() will be conflict.
My PR gets port based on IP and Name of instance.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Fix #43909

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

FengyunPan commented:
/assign @brendandburns

dims commented:
/ok-to-test

dims commented:
@FengyunPan why do we need new gophercloud? (what bug fixes or features do we need?)

FengyunPan commented:
@dims My PR can fix #43909. Please take a look. Some detail message can be get from there.
I have added code into gophercloud to support "nova interface-list <instance-id>", so it is need to update gophercloud.
Now I get "FAILED   hack/make-rules/../../hack/verify-staging-godeps.sh	356s", I try to fix it.

dims commented:
@FengyunPan : ah great. can you please have a separate commit for the gophercloud deps change? so it's easier to review

FengyunPan commented:
@dims Yeah, I agree, there are two commits in my PR, the first commit is to update gophercloud deps and the second commit fixs that bug.

FengyunPan commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-gce-etcd3

zhouhaibing089 commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -444,6 +445,65 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
+	srv, err := getServerByName(client, name)
+	if err != nil {
+		if err == ErrNotFound {
+			return interfaces, cloudprovider.InstanceNotFound
```
I know such type check happened at somewhere else, but wonder why do we have to define `ErrNotFound` rather than use `cloudprovider.InstanceNotFound1 directly?

FengyunPan replied:
I also think cloudprovider.InstanceNotFound is better than ErrNotFound.
In order to be easier to review, I will new PR to update ErrNotFound which comes from getServerByName().
Thank for your comment. @zhouhaibing089

FengyunPan replied:
Ok, that make sense.

zhouhaibing089 replied:
@anguslees my point is: it is really confusing that a caller relies on the error type(or value?) to do different things:

1. as an implementer, I have to know how caller will invoke my code.
2. if the caller really cares whether the server exists or not, it is better to be explicit as a return value, and no error should be returned.

for example, `getInterfacesByName` returns an empty list along with a NotFoundError, what does this mean, is it really an error, does an empty list already imply NotFound?

I am not talking about the code here specifically though, it is a general idea.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -246,7 +247,7 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
 	defer onFailure.Call(unwind)
 
-	port, err := getPortByIP(r.network, addr)
+	port, err := getPortByIP(r.compute, r.network, route.TargetNode, addr)
```
`updateAllowedAddressPairs` only actually needs the port ID .. so we should change that function to accept port ID (as a string) directly, and then adjust the code here to not bother fetching the full port object.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -276,3 +277,41 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
+	return port, nil
+}
+
+func getPortByIP(compute *gophercloud.ServiceClient, network *gophercloud.ServiceClient,
```
This is now a rather deep function, that (eg) does openstack lookups several levels into the callstack.  From experience, this is going to lead to duplicate lookups for things like `getServerByName`, additional code for slightly different variations, etc.

Stepping back, we want to do some rather simple things here:
- Find the ports on `route.TargetNode`
- Find which one of those ports has the address `addr`

I think it's reasonable to expose the individual steps here rather than hide them within nested functions so we can reuse those building blocks efficiently elsewhere.  So build reusable pieces for the server lookups, and then change `getPortByIP` into a local "shallow" helper that only exists to avoid repeating ourselves across the Create+Delete cases above - something like:
```go
func getPortIDByIP(...) {
   // skipped error handling for clarity, but you get the idea
   srv := getServerByName(compute, targetNode)
   interfaces := getAttachedInterfacesByID(compute, srv.ID)
   var portID string
   for _, intf := range interfaces {
      for _, fixedIP : = range intf.FixedIPs {
         if fixedIP.IPAddress == ipAddress {
            portID = intf.PortID
            break
         }
      }
   }
   if portID == "" {
      return ErrNotFound
   }
   return portID
}
```

FengyunPan replied:
Nice comment, thank you very much, I update it soon.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -444,6 +445,65 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
+	var interfaces []attachinterfaces.Interface
+
+	// get server
+	srv, err := getServerByName(client, name)
```
This function combines two server lookups (`getServerByName` and `attachinterfaces.List`).  Simplify it to only do one (the latter), and take the server ID as a parameter (rather than `NodeName`) - this will make it easier to compose with other functions later.  (ie: change this to `getAttachedInterfacesByID`)

FengyunPan replied:
Done.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -444,6 +445,65 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
+}
+
+// NodeFixedIPToPorts returns the node fixedIPs and ports of the specified instance.
+func getFixedIPToPortsByName(client *gophercloud.ServiceClient, name types.NodeName) (map[string]string, error) {
```
This function combines a server lookup, and some arbitrary logic.  It will be easier to test/reuse if it just does one of those two things, and we already have the server lookup function.   So make this just do the arbitrary logic - ie something like `findPortWithFixedIP([]attachinterfaces.Interface, addr string) (string, error)` - or else just drop the function and inline this logic into the caller (since we only want it in one place afaics).

anguslees reviewed (changes requested):
I like the general implementation.  I'm suggesting a bunch of refactoring to make the pieces more easily reusable in future.

FengyunPan commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -276,3 +277,41 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
+		return port, ErrNotFound
+	}
+
+	port = *targetPort
```
This function will be called by CreateRoute and DeleteRoute, we can collect these code into a func().

FengyunPan commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-bazel

dims commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -444,6 +445,65 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
+	if len(interfaceList) == 0 {
+		return interfaces, ErrNotFound
+	}
+	return interfaces, nil
```
@FengyunPan You are going to fix this right?

FengyunPan replied:
@dims Yes, I update my PR to make it simpler, and have removed interfaceList.
Please add comment into latest PR if you have any questions, thank you.

dims commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -276,3 +286,44 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
+	for _, intf := range interfaces {
+		for _, fixedIP := range intf.FixedIPs {
+			if fixedIP.IPAddress == ipAddress {
+				portID = intf.PortID
```
Can we just return from here? `return intf.PortID, nil`

dims replied:
since the break does not really break out of both the for loops

FengyunPan replied:
Yeah, that's ok.

dims commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -276,3 +286,44 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
+		}
+	}
+
+	if portID == "" {
```
If we do the above, we can skip the if condition here and just `return "", ErrNotFound` (and remove line 315 too)

FengyunPan replied:
Done.

dims commented:
/approve

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -444,6 +445,30 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
+	}
+
+	// return ErrNotFound when server contains no interface
+	if len(interfaces) == 0 {
```
As noticed in side discussion, this function is _intended_ to return a list, and so an empty list is not an error.

We should remove this `if` block.

FengyunPan replied:
Done.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_routes.go:
```diff
@@ -195,7 +200,7 @@ func (r *Routes) CreateRoute(clusterName string, nameHint string, route *cloudpr
 			IPAddress: route.DestinationCIDR,
 		})
-		unwind, err := updateAllowedAddressPairs(r.network, &port, newPairs)
+		unwind, err := updateAllowedAddressPairs(r.network, port, newPairs)
```
Change `updateAllowedAddressPairs` to take the port ID (rather than full Port object).  Then we can remove the `getPortByID()` call above.

FengyunPan replied:
It is need to check the AllowedAddressPairs of port before calling updateAllowedAddressPairs.
And updateAllowedAddressPairs will call port.AllowedAddressPairs too.

anguslees reviewed (changes requested):
Thanks for your persistence @FengyunPan - I only have one remaining minor request to remove an unnecessary lookup.  Almost there!

dims commented:
/approve

dims commented:
@brendanburns Can you please approve/lgtm? Thanks!

FengyunPan commented:
Erh, It seems not active.
/assign @thockin  Can you please approve/lgtm? Thanks!

FengyunPan commented:
Erh, he seems offline.
/assign @thockin 
Can you please approve/lgtm? Thanks!

FengyunPan commented:
/assign @smarterclayton 
@smarterclayton Can you please approve/lgtm? Thanks!

anguslees commented:
/lgtm

dims commented:
/retest

FengyunPan commented:
@dims It need a rebase. I will do a rebase soon.

Commit:
Update gophercloud to support list interfaces of OpenStack instance

Commit:
Fix conflict about getPortByIp

Currently getPortByIp() get port of instance only based on IP.
If there are two instances in diffent network and the CIDR of
their subnet are same, getPortByIp() will be conflict.
My PR gets port based on IP and Name of instance.

FengyunPan commented:
/test pull-kubernetes-federation-e2e-gce
/test pull-kubernetes-e2e-gce-etcd3

FengyunPan commented:
/test pull-kubernetes-federation-e2e-gce

FengyunPan commented:
Can you please add approve/lgtm label? @anguslees @brendandburns @dims @smarterclayton @thockin

dims commented:
@FengyunPan if i add lgtm, it won't merge since approvals are not complete :(

brendandburns commented:
/approve no-issue
/lgtm
